### PR TITLE
[RegAllocEvictAdvisor] Add minimum weight ratio heuristic.

### PR DESCRIPTION
**Pull request by mgudim**
Do not evict a live range if eviction will break existing hint without satisfying a new one and the ratio of weights is not big enough.

### CODE REVIEW

File: llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp [hunk 1]
```diff
@@ -44,6 +44,12 @@ static cl::opt<bool> EnableLocalReassignment(
              "may be compile time intensive"),
     cl::init(false));
 
+static cl::opt<float> MinWeightRatioNeededToEvictHint(
+    "min-weight-ratio-needed-to-evict-hint", cl::Hidden,
+    cl::desc("The minimum ration of weights needed in order for the live range with bigger weight to evict the other live range which"
+    "satisfies a hint),
+    cl::init(7.5));
```
Comment (mtrofin) [inline llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:51]: This should default to the value that makes this change a NFC, so `-inf`?

File: llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp [hunk 2]
```diff
@@ -157,6 +163,10 @@ bool DefaultEvictionAdvisor::shouldEvict(const LiveInterval &A, bool IsHint,
     return true;
 
   if (A.weight() > B.weight()) {
+    float WeightRatio = A.weight() / B.weight();
```
Comment (mtrofin) [inline llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:166]: can `B.weight()` be 0?

**Review by mtrofin — COMMENTED** [review]
Shouldn't there be a test?

File: llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp (hunk 1 above)
Comment (qcolombet) [inline llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:51]: To add on what @mtrofin said, could you explain the rationale for `7.5`?

It feels random to me.

**Review by qcolombet — CHANGES_REQUESTED** [review]

File: llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp [hunk 3]
```diff
@@ -44,6 +44,12 @@ static cl::opt<bool> EnableLocalReassignment(
              "may be compile time intensive"),
     cl::init(false));
 
+static cl::opt<float> MinWeightRatioNeededToEvictHint(
+    "min-weight-ratio-needed-to-evict-hint", cl::Hidden,
+    cl::desc("The minimum ration of weights needed in order for the live range with bigger weight to evict the other live range which"
```
[the remark this answers is not in the document]
Comment (mgudim) [reply llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp]: Fixed, thanks!

**Comment by mgudim** [comment]
> Shouldn't there be a test?

Added test

File: llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp (hunk 2 above)
Comment (mgudim) [reply llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp]: fixed

**Comment by mgudim** [comment]
@mtrofin @qcolombet 

This patch makes a lot of difference only on top of this: https://github.com/llvm/llvm-project/pull/90819
It actually gives good improvements in terms of dynamic instruction count on gcc and perlbench benchmarks.
I choose the number `7.5` by experimenting on xz where I originally had a degradation, but with 7.5 it's actually a slight improvement. Turned out the same `7.5` worked pretty well for the rest of the spec benchmarks. Perhaps a better number exists, but this needs a lot of experimentation. Maybe we should make the number architecture-dependent or even set it programatically depending on the code compiled? For now the value of `1.0` disables the heuristic by default.

**Comment by mgudim** [comment]
I've looked at the CI error - it is some assertion triggered from SLP vectorizer while compiling something in libcxx. I can't imagine how this change could trigger this, so I suspect something in the upstream is broken.

**Comment by mgudim** [comment]
@mtrofin @qcolombet After rebasing, the CI passes. Is this good to merge?

File: llvm/test/CodeGen/RISCV/min-weight-ratio.mir [hunk 4]
```diff
@@ -0,0 +1,76 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc %s -mtriple=riscv64 -run-pass=greedy,virtregrewriter \
+# RUN: -min-weight-ratio-needed-to-evict-hint=7.5 -o - | FileCheck %s
```
Comment (qcolombet) [inline llvm/test/CodeGen/RISCV/min-weight-ratio.mir:3]: To go back to what I was saying wrt 7.5. I'm not questioning that this gives you the best outcome in what you measured, what I was asking is why it is.
Essentially what is the explanation that makes 7.5 a good number aside from "it fixes my problem".

**Comment by mgudim** [comment]
the problem with the original commit was that it was hard to explain why it works. @qcolombet  what about this? : https://github.com/llvm/llvm-project/pull/98109/changes/ac67a96a42eb74fbb90db8287efb43b187fd4407 

On top of main, this commit doesn't really have much impact on spec.
On top of the early-csrs work it doest a little better than the original min-weight-ratio.

**Comment by mgudim** [comment]
> the problem with the original commit was that it was hard to explain why it works. @qcolombet what about this? : [ac67a96](https://github.com/llvm/llvm-project/commit/ac67a96a42eb74fbb90db8287efb43b187fd4407)
> 
> On top of main, this commit doesn't really have much impact on spec. On top of the early-csrs work it doest a little better than the original min-weight-ratio.

ping @qcolombet During your office hours you asked to ping you again to bring this up in your inbox

**Comment by mgudim** [comment]
@qcolombet ping

**Comment by mgudim** [comment]
ping @qcolombet

File: llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp
Comment (qcolombet) [inline llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:51]:
```suggestion
    cl::desc("Make eviction decisions based on simple weight comparison"),
```

Comment (qcolombet) [reply llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:51]: Don't know if it will become clearer later in the PR yet, but we should explain somewhere what "simple weight" means.

Comment (qcolombet) [inline llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:57]: What does "use-def frequency" mean?
It this the cumulated frequencies of all of A's uses and defs?

Comment (qcolombet) [inline llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:59]: The range seems a bit random to me.
Can you explain the rationale behind `1/MinUseDefFreqRatioToIgnoreSize`?
Like for the upper bound, what you're saying is if A is used/defined more than 5x than B, we ignore how much the live-range spans across the program (aka its size).

Comment (qcolombet) [inline llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:60]:
```suggestion
        "ignore the sizes of A and B when making the eviction decision."),
```

**Review by qcolombet — COMMENTED** [review]
My general comment didn't change.
I'm not questioning that you're getting better results in your benchmark, but I still would like to understand why this is the case.
These numbers seem still too magic to me.

File: llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp
Comment (mgudim) [reply llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:57]: yes

Comment (mgudim) [reply llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:51]: Here "simple" is attached to "comparison", not "weight". I wanted to say "we simply compare the weights of A and B" as opposed to the new heuristic

Comment (mgudim) [reply llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:59]: yes, exactly. I.e. the same upper bound definition is symmetric wrt A and B.

**Comment by mgudim** [comment]
> My general comment didn't change. I'm not questioning that you're getting better results in your benchmark, but I still would like to understand why this is the case. These numbers seem still too magic to me.

Before we were comparing weights of A and B. Weight is basically (use-def frequency) / size, so it treats size and use-def frequency as equally important. As it turns out, this does not work so well when there are a lot of huge live ranges with few uses. Under the "simple weight comparison" these can  be evicted too easily. For example, some cold range can evict it, while it shouldn't. At the same time we want these ranges to be evictable by ranges with high use-def frequency. This was my motivation to try to separate size and use-def frequency and let the latter be the main deciding factor. Since we can't  know how much more important use-def frequency is, I let it be a parameter, which I tuned based on the benchmarks. Does this give any intuition?

**Comment by qcolombet** [comment]
> > My general comment didn't change. I'm not questioning that you're getting better results in your benchmark, but I still would like to understand why this is the case. These numbers seem still too magic to me.
> 
> Before we were comparing weights of A and B. Weight is basically (use-def frequency) / size, so it treats size and use-def frequency as equally important. As it turns out, this does not work so well when there are a lot of huge live ranges with few uses. Under the "simple weight comparison" these can be evicted too easily. For example, some cold range can evict it, while it shouldn't. At the same time we want these ranges to be evictable by ranges with high use-def frequency. This was my motivation to try to separate size and use-def frequency and let the latter be the main deciding factor. Since we can't know how much more important use-def frequency is, I let it be a parameter, which I tuned based on the benchmarks. Does this give any intuition?

Thanks for that. I get the intuition now.

Let me step back to make sure we're on the same page around the rational of how the weight is computed.
The `(use-def frequency) / size` is supposed to answer the trade-off: how expensive this live-range is to spill vs. how much does it help other live-ranges.
The `size` is a proxy for "how much it helps other live-ranges" part and maybe what you're seeing is that this proxy fails in your case.

> As it turns out, this does not work so well when there are a lot of huge live ranges with few uses. Under the "simple weight comparison" these can be evicted too easily.

I think that what you are describing is ranges that end up with very close weights and the size aspect is kind of an arbitrary tie breaker.
Could you maybe share a debug log (`debug-only regalloc`) of a problematic case and the same debug log with your patch?
Even better if you can share a reproducer!

Bottom line, I still think this is the wrong thing to do. The fact that these numbers work well in your case doesn't mean they're going to be bad in other cases. My point is, with all heuristics being equally good/bad (as in on average they don't move the needle), I'd rather settle on simpler code, i.e., not having this additional layer of heuristic.
